lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -1095,31 +1199,111 @@ void ADContext::emitNondifferentiabilityError(SILInstruction *inst,
+    return nullptr;
+  auto *nominal = type->getAnyNominal();
+  assert(nominal && "Should've been nominal since it conforms to protocols");
+  return nominal;
+}
+
+/// Determines the cotangent space of a type.
+Optional<CotangentSpace> ADContext::getTangentSpace(CanType type) const {
+  LLVM_DEBUG(getADDebugStream() << "Classifying cotangent space for "
+             << type << '\n');
+  auto lookup = cachedTangentSpaces.find(type);
+  if (lookup != cachedTangentSpaces.end())
+    return lookup->getSecond();
+  // A helper that is used to cache the computed cotangent space for the specified
+  // type and retuns the same cotangent space.
+  auto cache = [&](Optional<CotangentSpace> cotangentSpace) {
+    cachedTangentSpaces.insert({type, cotangentSpace});
+    return cotangentSpace;
+  };
+  // `Builtin.FP<...>` is a builtin real scalar space.
+  if (auto *fpType = type->getAs<BuiltinFloatType>())
+    return cache(CotangentSpace::getBuiltinRealScalarSpace(fpType));
+  // Types that conform to `FloatingPoint` are a real scalar space.
+  if (auto *nomTy = getAnyRealScalarTypeDecl(type, *this))
+    return cache(CotangentSpace::getRealScalarSpace(nomTy));
+  // Types that conform to `VectorNumeric` where the associated `ScalarElement`
+  // conforms to `FloatingPoint` are a real vector space.
+  if (auto *nomTy = getAnyRealVectorTypeDecl(type, *this))
+    return cache(CotangentSpace::getRealVectorSpace(nomTy));
+  // Nominal types can be either a struct or an enum.
+  if (auto *nominal = type->getAnyNominal()) {
+    // Fixed-layout struct types, each of whose elements has a cotangent space,
+    // are a product of those cotangent spaces.
+    if (auto *structDecl = dyn_cast<StructDecl>(nominal)) {
+      if (structDecl->getFormalAccess() >= AccessLevel::Public &&
+          !structDecl->getAttrs().hasAttribute<FixedLayoutAttr>())
+        return cache(None);
+      auto allMembersHaveTangentSpace =
+          llvm::all_of(structDecl->getStoredProperties(), [&](VarDecl *v) {
+            return (bool)getTangentSpace(v->getType()->getCanonicalType());
+          });
+      if (allMembersHaveTangentSpace)
+        return cache(CotangentSpace::getProductStruct(structDecl));
+    }
+    // Frozen enum types, all of whose payloads have a cotangent space, are a sum
+    // of the product of payloads in each case.
+    if (auto *enumDecl = dyn_cast<EnumDecl>(nominal)) {
+      if (enumDecl->getFormalAccess() >= AccessLevel::Public &&
+          !enumDecl->getAttrs().hasAttribute<FrozenAttr>())
+        return None;
+      auto allMembersHaveTangentSpace =
+        llvm::all_of(enumDecl->getAllCases(), [&](EnumCaseDecl *cd) {
+          return llvm::all_of(cd->getElements(), [&](EnumElementDecl *eed) {
+            return llvm::all_of(*eed->getParameterList(), [&](ParamDecl *pd) {
+              return (bool)
+                  getTangentSpace(pd->getType()->getCanonicalType());
```

Comment:
can an indirect enum that refers to itself make this an infinite recursion?

Reply:
Good point, this should probably split on "indirect" edges.

Reply:
Thanks. Done.